File: 467 Unique Substrings in Wraparound String/467 Unique Substrings in Wraparound String.py
"""
We define the string s to be the infinite wraparound string of "abcdefghijklmnopqrstuvwxyz", so s will look like this:

    "...zabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcd....".

Given a string p, return the number of unique non-empty substrings of p are present in s.

Example 1:

Input: p = "a"
Output: 1
Explanation: Only the substring "a" of p is in s.

Example 2:

Input: p = "cac"
Output: 2
Explanation: There are two substrings ("a", "c") of p in s.

Example 3:

Input: p = "zab"
Output: 6
Explanation: There are six substrings ("z", "a", "b", "za", "ab", and "zab") of p in s.
"""

# dp[i] = the length of the longest conforming substrings ending at p[i]
# the key is the UNIQUE of the string
# for each substring, the ending char of the substring and the length uniquely determine the substring
# also, notice that for each index i in p, if there is a conforming substring ending at p[i] of length L, then there must also be substrings of length L-1, L-2, ...1.
# so for each p[i], the length of longest conforming substring is also the number of conforming substring.
# for each p[i], we maintain the length of the longest conforming substring
# also, we need to know the max length of conforming for each letter in A
# edge case: "cdefghefghi"

class Solution:
    def findSubstringInWraproundString(self, p: str) -> int:
        A = string.ascii_lowercase
        pre = {A[i]:A[i-1] for i in range(1,len(A))}
        pre['a'] = 'z'

        total = 1 # for p[0]
        ML = {c:0 for c in A} # max conforming length for each letter in A we've seen currently
        ML[p[0]] = 1 # we start for loop from p[1], so we need to manually tackle ML[p[0]]
        dp = [1] * len(p) # max conforming length for each index in p

        for i in range(1,len(p)):
            if pre[p[i]] == p[i-1]:
                dp[i] = 1 + dp[i-1]
            if dp[i] > ML[p[i]]:
                total += dp[i] - ML[p[i]]
                ML[p[i]] = dp[i]
        return total

File: 467 Unique Substrings in Wraparound String/467 Unique Substrings in Wraparound String.go
func findSubstringInWraproundString(p string) int {
    if len(p) == 0 {
        return 0
    }
    pre := 1
    res := 0
    maxLen := make([]int, 26) //use maxLen slice to maintain the max len of substring of each letter
    maxLen[p[0] - 'a'] = 1

    for i := 1; i < len(p); i++ {
        if p[i] - p[i - 1] == 1 || p[i - 1] - p[i] == 25 {
            pre++ //if current letter is at alphabet sequence of previous letter, increase 1 to the max len
        } else {
            pre = 1 //otherwise the max len is 1
        }
        maxLen[p[i] - 'a'] = max(maxLen[p[i] - 'a'], pre) //update the maxLen of this letter
    }
    for _, v := range maxLen {
        res += v //sum up the max len of each letters
    }
    return res
}

func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}
